### PR TITLE
Fix/remotelock service to lockservice

### PR DESCRIPTION
**Pull request by hsaraogi**
**Goals (and why)**: Expose the lock service API so that the large internal product can use Timelock Service

**Implementation Description (bullets)**:
1. Expose the lock service API
2. Correnctly implement serde for some return types for the Lock API.
3. Tests

**Concerns (what feedback would you like?)**: Is it safe to expose the lock service to all users? I think it is, given we are not exposing it on timelock and we expect most users to switch to timelock.

**Where should we start reviewing?**: TimelockServerLauncher

**Priority (whenever / two weeks / yesterday)**: one week

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2284)

### CODE REVIEW

File: timelock-server-distribution/build.gradle [hunk 1]
```diff
@@ -16,3 +16,10 @@ distribution {
     args 'server', 'var/conf/timelock.yml'
     defaultJvmOpts "-Xms512m", "-Xmx512m", "-javaagent:service/lib/jetty-alpn-agent-${libVersions.jetty_alpn_agent}.jar"
 }
+
+task runTimeLock(type: JavaExec) {
```
Comment (hsaraogi) [inline timelock-server-distribution/build.gradle]: This was added to run timelock service locally without intellij.

**Comment by datval** [comment]
Do we need to check in timelock-server-distribution/service/lib/jetty-alpn-agent-2.0.6.jar ?

**Comment by hsaraogi** [comment]
@datval it is checked in, just makes it easier to launch the server with HTTP/2 and does not cause any harm I think.

**Comment by fsamuel-bs** [comment]
test

**Comment by hsaraogi** [comment]
hey!

File: lock-api/src/main/java/com/palantir/lock/LockResponse.java [hunk 2]
```diff
@@ -103,10 +114,19 @@ public boolean success() {
...
     @Nullable public LockRefreshToken getLockRefreshToken() {
         return token == null ? null : token.getLockRefreshToken();
     }
 
+    public boolean isBlockAndRelease() {
+        return isBlockAndRelease;
+    }
+
+    public List<LockWithClient> getLocks() {
+        return ImmutableList.copyOf(Iterables.transform(lockHolders.entrySet(), TO_LOCK_WITH_CLIENT_FUNCTION));
```
Comment (nziebart) [inline lock-api/src/main/java/com/palantir/lock/LockResponse.java]: nit: the function could probably just be inline here

File: lock-api/src/main/java/com/palantir/lock/LockServerOptions.java [hunk 3]
```diff
@@ -180,32 +206,43 @@ public String getLockStateLoggerDir() {
...
-                    return maxNormalLockAge;
-                }
-                @Override public int getRandomBitCount() {
-                    return randomBitCount;
-                }
-                @Override public String getLockStateLoggerDir() {
-                    return lockStateLoggerDir;
-                }
-            };
+            return new Builder()
```
Comment (nziebart) [inline lock-api/src/main/java/com/palantir/lock/LockServerOptions.java]: nice, glad to see this cleaned up

File: lock-api/src/main/java/com/palantir/lock/LockService.java [hunk 4]
```diff
@@ -87,6 +107,10 @@
      *         {@code token} are read locks (instead of write locks), or if
      *         {@code token} is held by an anonymous lock client
      */
+    @POST
+    @Path("unlock-and-freeze")
```
Comment (nziebart) [inline lock-api/src/main/java/com/palantir/lock/LockService.java]: i assume these names are all taken from PG's lock server?

File: lock-api/src/main/java/com/palantir/lock/LockWithClient.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.lock;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class LockWithClient {
```
Comment (nziebart) [inline lock-api/src/main/java/com/palantir/lock/LockWithClient.java]: can this be implemented with immutables?

File: lock-impl/src/main/java/com/palantir/lock/impl/ThreadPooledLockService.java [hunk 6]
```diff
@@ -64,6 +70,84 @@ public Long getMinLockedInVersionId(String client) {
     }
 
     @Override
+    public LockResponse lockWithFullLockResponse(LockClient client, LockRequest request) throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (nziebart) [reply lock-impl/src/main/java/com/palantir/lock/impl/ThreadPooledLockService.java]: all the `lock` methods should be limited (i.e., the methods that can block); anything else doesn't need to be

File: timelock-server-distribution/build.gradle [hunk 7]
```diff
@@ -16,3 +16,10 @@ distribution {
...
+    classpath sourceSets.main.runtimeClasspath
+    main = "com.palantir.atlasdb.timelock.TimeLockServerLauncher"
+    jvmArgs = ["-javaagent:service/lib/jetty-alpn-agent-2.0.6.jar"]
+    args "server", "var/conf/timelock2.yml"
```
Comment (nziebart) [inline timelock-server-distribution/build.gradle]: should we name this something better? any reason it can't be `timelock.yml`?

**Comment by nziebart** [comment]
@datval no; we'll need to add a config option to use the kvs timestamp bound store and wire it up

**Comment by jeremyk-91** [comment]
(For SA - I'll have a look at this at some point today as well.)

File: docs/source/release_notes/release-notes.rst [hunk 8]
```diff
@@ -44,6 +44,15 @@ develop
     *    - Type
          - Change
 
+    *    - |devbreak|
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: If I understand the PR correctly, this is still possible though, so it's not really a dev break.

File: lock-api/src/main/java/com/palantir/lock/LockResponse.java [hunk 9]
```diff
@@ -88,6 +98,7 @@ public static LockResponse createSuccessful(HeldLocksToken token) {
      * behavior) or {@link BlockingMode#BLOCK_INDEFINITELY_THEN_RELEASE},
      * then this method is guaranteed to return <code>true</code>.
      */
+    @JsonIgnore
```
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/LockResponse.java]: nit: I don't think this is needed, as it's not a getter.

File: lock-api/src/main/java/com/palantir/lock/LockService.java (hunk 4 above)
Comment (jeremyk-91) [reply lock-api/src/main/java/com/palantir/lock/LockService.java]: If I'm not wrong the original doesn't use a rest-based framework, so there isn't really a precedent. The protobuf lock service had an enum for the various methods, as we would probably do if we did a websocket like impl.

File: lock-api/src/main/java/com/palantir/lock/LockResponse.java (hunk 2 above)
Comment (hsaraogi) [reply lock-api/src/main/java/com/palantir/lock/LockResponse.java]: Done, replaced with stream.

File: lock-api/src/main/java/com/palantir/lock/LockService.java (hunk 4 above)
Comment (hsaraogi) [reply lock-api/src/main/java/com/palantir/lock/LockService.java]: Yes, as PG does not have REST APIs, I came up with these names based on the method names.

File: lock-api/src/main/java/com/palantir/lock/LockWithClient.java (hunk 5 above)
Comment (hsaraogi) [reply lock-api/src/main/java/com/palantir/lock/LockWithClient.java]: Done.

File: timelock-server-distribution/build.gradle (hunk 7 above)
Comment (hsaraogi) [reply timelock-server-distribution/build.gradle]: timelock.yml does not expose the HTTP/2 connector. This task is aimed at making it easier to test against the large internal product. Renaming it to `timelock-http2.yml`.

File: docs/source/release_notes/release-notes.rst (hunk 8 above)
Comment (hsaraogi) [reply docs/source/release_notes/release-notes.rst]: true.

File: lock-api/src/main/java/com/palantir/lock/LockClient.java [hunk 10]
```diff
@@ -88,9 +87,7 @@ public String getClientId() {
     }
 
     @Override public String toString() {
-        return MoreObjects.toStringHelper(getClass().getSimpleName())
-                .addValue(isAnonymous() ? "anonymous" : ("\"" + clientId + "\""))
-                .toString();
+        return clientId;
```
Comment (hsaraogi) [inline lock-api/src/main/java/com/palantir/lock/LockClient.java]: Is this break concerning?

Comment (nziebart) [reply lock-api/src/main/java/com/palantir/lock/LockClient.java]: This break seems fine. No ones should have logic depending on the toString impl

File: timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceIntegrationTest.java [hunk 11]
```diff
@@ -190,6 +206,106 @@ public void canGetCurrentTimeMillis() {
...
+        cluster.lockService().useGrant(TEST_CLIENT_2, heldLocksGrant);
+        assertThat(Iterables.getOnlyElement(cluster.lockService().getTokens(TEST_CLIENT_2)).getLockDescriptors())
+                .contains(LOCK_A);
+
+        assertThatThrownBy(() -> cluster.lockService().useGrant(TEST_CLIENT_3, heldLocksGrant.getGrantId()))
+                .isInstanceOf(AtlasDbRemoteException.class);
+    }
+
+    @Test
+    public void getMinLockedInVersionIdWorks() throws InterruptedException {
```
Comment (nziebart) [inline timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceIntegrationTest.java]: Could this have a more descriptive name? Looks like its checking the behavior when there is no min version id

File: timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceIntegrationTest.java [hunk 12]
```diff
@@ -190,6 +206,106 @@ public void canGetCurrentTimeMillis() {
...
+    public void canPerformLockAndUnlock() throws InterruptedException {
+        HeldLocksToken token1 = lockWithFullResponse(requestForWriteLock(LOCK_A), TEST_CLIENT);
+        cluster.lockService().unlock(token1);
+        HeldLocksToken token2 = lockWithFullResponse(requestForWriteLock(LOCK_A), TEST_CLIENT);
+        cluster.lockService().unlockSimple(SimpleHeldLocksToken.fromHeldLocksToken(token2));
+    }
+
+    @Test
+    public void canPerformLockAndRefreshLock() throws InterruptedException {
+        lockWithFullResponse(requestForWriteLock(LOCK_A), TEST_CLIENT);
```
Comment (nziebart) [inline timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceIntegrationTest.java]: Is this first line needed?

File: lock-api/src/main/java/com/palantir/lock/LockClient.java (hunk 10 above)
Comment (jeremyk-91) [reply lock-api/src/main/java/com/palantir/lock/LockClient.java]: Yep. I'm not sure I like `LockClient.ANONYMOUS` being a completely empty string, but don't think we should block on that.

**Comment by hsaraogi** [comment]
@jeremyk-91 @nziebart Thanks for approving, but some ETE tests were failing and I found a bug and fixed it with some more cleanup post the approval. Could you take a final look again before I merge this?

**Pull request merged**